app/models/concerns/thredded/topic_common.rb

```diff
@@ -47,24 +47,45 @@ def unread(user)
           .merge(reads_class.where(reads[:id].eq(nil).or(reads[:read_at].lt(topics[:updated_at]))))
       end
 
+      private
+
+      # @param user [Thredded.user_class]
+      # @return [ByPostableLookup]
+      def read_states_by_topics_lookup(user, null_value)
+        read_states = reflect_on_association(:user_read_states).klass
+          .where(user_id: user.id, postable_id: current_scope.map(&:id))
+        Thredded::TopicCommon::Lookup.new(read_states, null_value)
```

Comment:
@jayroh @glebm On reflection the utility class is doing two things -- managing the inverse lookup performance gain and also dealing with null values. If it was just dealing with the inverse lookup and was basically a hash, then it would make it a lot easier to understand and we wouldn't need so much code or documentation or specs. Am pushing a commit to show this